jetty-server/src/main/java/org/eclipse/jetty/server/DetectorConnectionFactory.java

```diff
@@ -0,0 +1,293 @@
+ * the first one recognizing the bytes in the buffer.
+ */
+public class DetectorConnectionFactory extends AbstractConnectionFactory
+{
+    private static final Logger LOG = Log.getLogger(DetectorConnectionFactory.class);
+
+    protected final List<Detecting> _detectingConnectionFactories;
+
+    /**
+     * <p>When the first bytes are not recognized by the {@code detectingConnectionFactories}, the default behavior is to
+     * upgrade to the protocol returned by {@link #findNextProtocol(Connector)}.</p>
+     * @param detectingConnectionFactories the {@link Detecting} instances.
+     */
+    public DetectorConnectionFactory(Detecting... detectingConnectionFactories)
+    {
+        super(toProtocolString(detectingConnectionFactories));
+        _detectingConnectionFactories = Arrays.asList(detectingConnectionFactories);
+        for (Detecting detectingConnectionFactory : detectingConnectionFactories)
+        {
+            addBean(detectingConnectionFactory);
+        }
+    }
+
+    static String toProtocolString(Detecting... detectingConnectionFactories)
+    {
+        if (detectingConnectionFactories.length == 0)
+            throw new IllegalStateException("At least one detecting instance is required");
+
+        // remove protocol duplicates while keeping their ordering -> use LinkedHashSet
+        LinkedHashSet<String> protocols = new LinkedHashSet<>();
+        for (Detecting dcf : detectingConnectionFactories)
+        {
+            protocols.add(dcf.getProtocol());
+        }
+
+        StringBuilder sb = new StringBuilder();
+        for (String protocol : protocols)
+        {
+            sb.append(protocol).append('|');
+        }
+        sb.deleteCharAt(sb.length() - 1);
+        String protocol = sb.toString();
+        LOG.debug("Detector generated protocol name : {}", protocol);
+        return protocol;
+    }
+
+    /**
+     * Perform a detection using multiple {@code detectingConnectionFactories} and return the aggregated outcome.
+     * @param buffer the buffer to perform a detection against.
+     * @param detectingConnectionFactories the {@link Detecting} instances.
+     * @return A {@link Detecting.Detection} value with the detection outcome of the {@code detectingConnectionFactories}.
+     */
+    protected static Detecting.Detection detectUsing(ByteBuffer buffer, Iterable<Detecting> detectingConnectionFactories)
+    {
+        LOG.debug("Attempting detection from buffer {} using {}", buffer, detectingConnectionFactories);
+        boolean allDetectionNotRecognized = true;
+        for (Detecting detectingConnectionFactory : detectingConnectionFactories)
+        {
+            Detecting.Detection detection = detectingConnectionFactory.detect(buffer);
+            if (detection == Detecting.Detection.RECOGNIZED)
+            {
+                LOG.debug("Detection recognized bytes from buffer {} using {}", buffer, detectingConnectionFactories);
+                return Detecting.Detection.RECOGNIZED;
```

Comment:
If more than one factory returns `RECOGNIZED`, should that instead be treated as `NEED_MORE_BYTES`?  This way an individual factory does not need to know how many bytes are required to disambiguate from other protocols.

Reply:
Currently, the same functionality can be achieved by cleverly ordering the detecting connection factories. Maybe that should be documented?

However, we could do what you're proposing as an improvement at the cost of a slightly more complex implementation, but I'm not certain it would be worth the effort, at least for now.

Reply:
I would still like to avoid the need to "cleverly order" factories if at all possible.  Is the extra code really that more complex?

Reply:
Well, thinking about it more I wonder what could we do in the case that more than one `Detecting` instance recognizes the bytes, besides eventually logging that?

Apart from troubleshooting conflicts between multiple `detect()` implementations, this wouldn't help, would it?